test/unit/sourceset.test.js

```diff
@@ -0,0 +1,532 @@
+      this.player = videojs(this.mediaEl);
+      this.player.src(this.testSrc);
+
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.testSrc]);
+        done();
+      });
+    });
+
+    QUnit.test('player.src({...}) two sources', function(assert) {
+      const done = assert.async();
+
+      this.player = videojs(this.mediaEl);
+      this.player.src([this.sourceOne, this.sourceTwo]);
+
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.sourceOne, this.sourceTwo]);
+        done();
+      });
+    });
+
+    QUnit.test('mediaEl.src = ...;', function(assert) {
+      const done = assert.async();
+
+      this.mediaEl.src = this.testSrc.src;
+      this.player = videojs(this.mediaEl);
+
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.testSrc]);
+        done();
+      });
+    });
+
+    QUnit.test('mediaEl.setAttribute("src", ...)"', function(assert) {
+      const done = assert.async();
+
+      this.mediaEl.setAttribute('src', this.testSrc.src);
+      this.player = videojs(this.mediaEl);
+
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.testSrc]);
+        done();
+      });
+    });
+
+    QUnit.test('<source> one source', function(assert) {
+      const done = assert.async();
+
+      this.source = document.createElement('source');
+      this.source.src = this.testSrc.src;
+      this.source.type = this.testSrc.type;
+
+      this.mediaEl.appendChild(this.source);
+
+      this.player = videojs(this.mediaEl);
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.testSrc]);
+        done();
+      });
+    });
+
+    QUnit.test('<source> two sources', function(assert) {
+      const done = assert.async();
+
+      this.source = document.createElement('source');
+      this.source.src = this.sourceOne.src;
+      this.source.type = this.sourceOne.type;
+
+      this.source2 = document.createElement('source');
+      this.source2.src = this.sourceTwo.src;
+      this.source2.type = this.sourceTwo.type;
+
+      this.mediaEl.appendChild(this.source);
+      this.mediaEl.appendChild(this.source2);
+
+      this.player = videojs(this.mediaEl);
+
+      this.player.one('sourceset', () => {
+        validateSource(assert, this.player, [this.sourceOne, this.sourceTwo]);
+        done();
+      });
+    });
+
+    QUnit.test('no source', function(assert) {
+      const done = assert.async();
+
+      this.player = videojs(this.mediaEl);
+
+      this.totalSourcesets = 0;
+
+      window.setTimeout(() => {
+        assert.equal(this.sourcesets, 0, 'no sourceset');
+        done();
+      }, wait);
+    });
+  }));
+
+  QUnit.module('source change', (subhooks) => testTypes.forEach((testName) => {
+    QUnit.module(testName, {
+      beforeEach(assert) {
+        sinon.stub(log, 'error');
```

Comment:
is there much difference in this `beforeEach` and the one above? Could probably pull out to one function above and reference both in the object here.

Reply:
I think we could pull it out.

Reply:
looks like there's a lot of overlap but there are some differences. I'll probably pull out the main bulk that's the same into a function.